src/lib.rs

```diff
@@ -121,18 +126,250 @@ impl Invariant {
+                    .expect("Attempted to remove incorrect tick");
+            }
+        };
+
+        let (mut position, x, y) = match Position::create(
+            &mut pool,
+            pool_key,
+            &mut lower_tick,
+            &mut upper_tick,
+            current_timestamp,
+            liquidity_delta,
+            slippage_limit_lower,
+            slippage_limit_upper,
+            current_block_number,
+            pool_key.fee_tier.tick_spacing,
+        ) {
+            Ok(position) => position,
+            Err(e) => {
+                undo_ticks_creation(
+                    self,
+                    lower_tick.liquidity_gross.is_zero(),
+                    lower_tick,
+                    upper_tick.liquidity_gross.is_zero(),
+                    upper_tick
+                );
+                return Err(e);
+            }
+        };
+
+        let undo_creating_position = |invariant: &mut Self,
+                                      position: &mut Position,
+                                      pool: &mut Pool,
+                                      mut lower_tick: Tick,
+                                      mut upper_tick: Tick,
+                                      pool_key: PoolKey| {
+            let (_x, _y, remove_lower_tick, remove_upper_tick) = position.remove(
+                pool,
+                current_timestamp,
+                &mut lower_tick,
+                &mut upper_tick,
+                pool_key.fee_tier.tick_spacing,
+            );
+
+            undo_ticks_creation(invariant, remove_lower_tick, lower_tick, remove_upper_tick, upper_tick);
+        };
+
+        let first_transaction = self
+            .transfer_tokens(&pool_key.token_x, None, &caller, &program, x.get())
+            .await;
+
+        if let Err(e) = first_transaction {
+            undo_creating_position(
+                self,
+                &mut position,
+                &mut pool,
+                lower_tick,
+                upper_tick,
+                pool_key,
+            );
+
+            return Err(e);
+        }
+        let second_transaction = self
+            .transfer_tokens(&pool_key.token_y, None, &caller, &program, y.get())
+            .await;
+
+        if let Err(e) = second_transaction {
+            undo_creating_position(
+                self,
+                &mut position,
+                &mut pool,
+                lower_tick,
+                upper_tick,
+                pool_key,
+            );
+            self.transfer_tokens(&pool_key.token_x, None, &program, &caller, x.get())
+                .await
+                .unwrap();
+
+            return Err(e);
+        }
+
+        self.pools.update(&pool_key, &pool)?;
+
+        self.positions.add(&caller, &position);
+
+        self.ticks.update(pool_key, lower_tick.index, lower_tick)?;
+        self.ticks.update(pool_key, upper_tick.index, upper_tick)?;
+     
+        self.emit_event(InvariantEvent::PositionCreatedEvent {
```

Comment:
Is this something we can check within e2e tests? If it is, it's worthwhile to also validate emitting events